Using split reduction to convert an inner-reduction into outer reduction + inner-parallel dimension, followed by simple inner reduction.

vmurali opened the pull request.

dcaballe commented:
Hey @pzread @hanhanW , do you know why benchmarks are not triggering for this PR? Should @vmurali be granted some extra permissions or something?

pzread commented:
> Hey @pzread @hanhanW , do you know why benchmarks are not triggering for this PR? Should @vmurali be granted some extra permissions or something?

Yeah... There is a weird issue on buildkite that I haven't had time to look. Invited @vmurali to our buildkite team as a work around.

vmurali commented:
Thanks Jerry; I don't think I have received the invite yet.

On Thu, Sep 29, 2022 at 10:37 PM Jerry Wu ***@***.***> wrote:

> Hey @pzread <https://github.com/pzread> @hanhanW
> <https://github.com/hanhanW> , do you know why benchmarks are not
> triggering for this PR? Should @vmurali <https://github.com/vmurali> be
> granted some extra permissions or something?
>
> Yeah... There is a weird issue on buildkite that I haven't had time to
> look. Invited @vmurali <https://github.com/vmurali> to our buildkite team
> as a work around.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/iree-org/iree/pull/10537#issuecomment-1263123662>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABZURPLYGCKOESKKJVU3GDWAZ37DANCNFSM6AAAAAAQU5G7JE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

vmurali commented:
Thanks @pzread

hanhanW commented:
I've seen that you keep pushing commits. Is the PR ready for review? If not, you can convert it to a draft PR.

<img width="478" alt="image" src="https://user-images.githubusercontent.com/6949191/193368613-49178fdc-29c2-43fc-b4e4-af3027a35278.png">

vmurali commented:
It was ready, but issues propped up. I have made it a draft for now.

MaheshRavishankar commented:
@vmurali broad comment before I look into details. Please add a flag to guard the split-reduction. It associatively reorders the computation (even if the guard is always true for now). There might be better ways to do this in the long term, but at least adding a flag for now would be useful.

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -138,6 +140,40 @@ getTileAndFuseOptionsFromConfig(func::FuncOp funcOp, int64_t tilingLevel) {
+  auto tileSizesAttr = configAttr.getTileSizes();
+  if (!tileSizesAttr || tileSizesAttr.empty()) return false;
+  // The last one is the reduction dimension.
+  auto reductionSizesAttr = tileSizesAttr[tileSizesAttr.size() - 1];
```
Probably better to explicitly pass in the position of the tile size tuple being used. Its an implementation detail that its the last tile level that is being used, but thats not always true. I think the position in the `tileSizesAttr` used should be an input to the pass itself so that it can be controlled from the pass pipeline.

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -138,6 +140,40 @@ getTileAndFuseOptionsFromConfig(func::FuncOp funcOp, int64_t tilingLevel) {
 
+/// Default method to initialize the split reduction size and unroll factor in
+/// IREE. These could be overriden by the command line options if specified.
+static bool getSplitReductionOptionsFromConfig(func::FuncOp funcOp,
```
Its easier to debug if you return a `FailureOr<...>` or `Optional<..>` . Return by reference is not the prefered coding style.

vmurali replied:
There are two return values.

vmurali replied:
I created a pair. That seems readable enough

dcaballe commented:
> > Hey @pzread @hanhanW , do you know why benchmarks are not triggering for this PR? Should @vmurali be granted some extra permissions or something?
> 
> Yeah... There is a weird issue on buildkite that I haven't had time to look. Invited @vmurali to our buildkite team as a work around.

Thanks! Is there a way to retrigger the benchmark pipelines? They don't seem to be running even after pushing new commits.

> @vmurali broad comment before I look into details. Please add a flag to guard the split-reduction. It associatively reorders the computation (even if the guard is always true for now). There might be better ways to do this in the long term, but at least adding a flag for now would be useful.

It could be the same as https://github.com/iree-org/iree/issues/10279

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -366,6 +411,128 @@ void LinalgFusePass::runOnOperation() {
   }
 }
+namespace {
+/// Pattern to wrap splitReduction transformation. This changes the
```
Is the comment still correct?

vmurali replied:
No

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -366,6 +411,128 @@ void LinalgFusePass::runOnOperation() {
+    rewriter.replaceOp(splitRes->splitLinalgOp, tileRes->tensorResults);
+    filter.replaceLinalgTransformationFilter(rewriter, tileRes->loops[0]);
+
+    // // 4) Unroll the tiled op (scf.for) by unrollFactor. loopUnrollByFactor
```
Remove commented code...

vmurali replied:
That code is an integral part of the PR, except that it crashes now (probably because `loopUnrollByFactor` is buggy). I would rather fix the bug in this PR if it's not too tedious.

hanhanW replied:
In this case, let's replace them with descriptive comments?

MaheshRavishankar reviewed (changes requested):
Could we move the cherry-pick into a separate commit?

vmurali commented:
> > Hey @pzread @hanhanW , do you know why benchmarks are not triggering for this PR? Should @vmurali be granted some extra permissions or something?
> 
> Yeah... There is a weird issue on buildkite that I haven't had time to look. Invited @vmurali to our buildkite team as a work around.

> Thanks @pzread

The benchmarks run now, but I am not able to see the results table.

vmurali commented on compiler/src/iree/compiler/Codegen/Sandbox/Passes.h:
```diff
@@ -38,6 +38,10 @@ std::unique_ptr<OperationPass<func::FuncOp>> createLinalgFusePass();
     const LinalgFusePassOptions &options);
 
+/// Creates a pass to perform SplitReduction transformations of `LinalgOp`s.
+std::unique_ptr<OperationPass<func::FuncOp>> createLinalgSplitReductionPass(
```
It's in the `Passes.td` file.

hanhanW commented:
> The benchmarks run now, but I am not able to see the results table.

It failed at build. Seems that there are compilation errors in some models: https://buildkite.com/iree/iree-benchmark-android/builds/4327#01838ce6-6a2f-4d8c-96be-9721c6307eb2

You can build models locally:

```
# Enable -DIREE_BUILD_BENCHMARKS in cmake
cmake src/ -DIREE_BUILD_BENCHMARKS 
cmake --build . --target iree-benchmark-suites
```

vmurali commented:
> > The benchmarks run now, but I am not able to see the results table.
> 
> It failed at build. Seems that there are compilation errors in some models: https://buildkite.com/iree/iree-benchmark-android/builds/4327#01838ce6-6a2f-4d8c-96be-9721c6307eb2
> 
> You can build models locally:
> 
> ```
> # Enable -DIREE_BUILD_BENCHMARKS in cmake
> cmake src/ -DIREE_BUILD_BENCHMARKS 
> cmake --build . --target iree-benchmark-suites
> ```

Where did you get that link from? I now see buildkite running with the following two links (both are pending): https://buildkite.com/iree/iree-benchmark-android/builds/4347 and https://buildkite.com/iree/iree-benchmark-linux/builds/1559

hanhanW commented:
> > > The benchmarks run now, but I am not able to see the results table.
> > 
> > 
> > It failed at build. Seems that there are compilation errors in some models: https://buildkite.com/iree/iree-benchmark-android/builds/4327#01838ce6-6a2f-4d8c-96be-9721c6307eb2
> > You can build models locally:
> > ```
> > # Enable -DIREE_BUILD_BENCHMARKS in cmake
> > cmake src/ -DIREE_BUILD_BENCHMARKS 
> > cmake --build . --target iree-benchmark-suites
> > ```
> 
> Where did you get that link from? I now see buildkite running with the following two links (both are pending): https://buildkite.com/iree/iree-benchmark-android/builds/4347 and https://buildkite.com/iree/iree-benchmark-linux/builds/1559

yeah, I re-triggered it because I did not see the benchmarks in checks.

You can click `Details` on

<img width="1232" alt="image" src="https://user-images.githubusercontent.com/6949191/193681065-2bf5f756-12c8-4861-b0ed-fcf12ce8dbb8.png">

And clicking the branch name (e.g., `vmurali:murali_iree`) leads you to previous results.
 
<img width="495" alt="image" src="https://user-images.githubusercontent.com/6949191/193681110-e980ba3c-5b88-446b-a376-732eb75af1a1.png">

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -366,6 +411,128 @@ void LinalgFusePass::runOnOperation() {
+    AffineMap map = op.getTiedIndexingMap(op.getInputOperand(0));
+    unsigned lastIdx = map.getNumResults() - 1;
+    unsigned lastDim = map.getDimPosition(lastIdx);
+    if (lastDim != dims[0])
+      return rewriter.notifyMatchFailure(op, "precondition not met");
```
style nit: We wrap the single statements with braces in IREE.

```suggestion
    if (lastDim != dims[0]) {
      return rewriter.notifyMatchFailure(op, "precondition not met");
    }
```

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -366,6 +411,128 @@ void LinalgFusePass::runOnOperation() {
+      return rewriter.notifyMatchFailure(op, "precondition not met");
+
+    linalg::ControlSplitReductionFn fn = [this, lastIdx](linalg::LinalgOp op) {
+      return linalg::SplitReductionOptions{size, lastIdx, true};
```
The option is not clear, let's add a function argument comment?

```suggestion
      return linalg::SplitReductionOptions{size, lastIdx, /*innerParallel=*/true};
```

https://google.github.io/styleguide/cppguide.html#Function_Argument_Comments

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -366,6 +411,128 @@ void LinalgFusePass::runOnOperation() {
+      return linalg::SplitReductionOptions{size, lastIdx, true};
+    };
+
+    auto opNumLoops = op.getNumLoops();
```
Maybe declare the variable before `lastIdx`, then we don't need `lastIdx` variable. It can be computed through `opNumLoops - 1`.

Also, I think it's not promising that the input operand contains all the dimensions in result map right?

[optional] Lets use `numLoops` like other naming in the file.

vmurali replied:
`opNumLoops` and `lastIdx` have nothing to do with each other. `opNumLoops` specifies the number of loops, and `lastIdx` is 1 less than the number of output dimensions.

> Also, I think it's not promising that the input operand contains all the dimensions in result map right?

Not sure what you mean here.

> [optional] Lets use numLoops like other naming in the file.
Fixed.

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -366,6 +411,128 @@ void LinalgFusePass::runOnOperation() {
+    auto opNumLoops = op.getNumLoops();
+
+    // 1) Tile to extract a single vector-length array.
+    SmallVector<int64_t> tileSizesSVFirst(opNumLoops, 1);
+    tileSizesSVFirst[opNumLoops - 1] = 0;
+    ArrayRef<int64_t> tileSizesFirst(tileSizesSVFirst);
+    linalg::LinalgTilingOptions optionsFirst;
+    optionsFirst.setTileSizes(tileSizesFirst);
```
We don't have to declare the ArrayRef type. The `SmallVector` will be casted to ArrayRef type in this case. Also, we can make setting options cleaner. I think the below snippet work, can you give it a shot?

```suggestion
    SmallVector<int64_t> tileSizesSVFirst(opNumLoops, 1);
    tileSizesSVFirst[opNumLoops - 1] = 0;
    auto optionsFirst = linalg::LinalgTilingOptions().setTileSizes(tileSizesSVFirst);
```

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -366,6 +411,128 @@ void LinalgFusePass::runOnOperation() {
+
+    // 3) Tile the first op generated by splitReduction with tile size of 1, to
+    // essentially create a reduction loop.
+    SmallVector<int64_t> tileSizesSV(opNumLoops + 1, 0);
+    tileSizesSV[opNumLoops - 1] = 1;
```
maybe should get the size from `splitRes->splitLinalgOp.getNumLoops`. This requires others to understand how splitReduction works. I'm not sure if it's always `+1`.

Also, why `tileSizesSV[opNumLoops - 1] = 1;`? can you add a comment on `opNumLoops - 1`?

vmurali replied:
> maybe should get the size from splitRes->splitLinalgOp.getNumLoops. This requires others to understand how splitReduction works. I'm not sure if it's always +1.

It's always `+1`. Added a comment.

Added a comment explaining why it's `numLoops - 1`

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -366,6 +411,128 @@ void LinalgFusePass::runOnOperation() {
+    // essentially create a reduction loop.
+    SmallVector<int64_t> tileSizesSV(opNumLoops + 1, 0);
+    tileSizesSV[opNumLoops - 1] = 1;
+    ArrayRef<int64_t> tileSizes(tileSizesSV);
+    linalg::LinalgTilingOptions options;
+    options.setTileSizes(tileSizes);
```
ditto: we don't need ArrayRef and can have better readability form for setting options.

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -366,6 +411,128 @@ void LinalgFusePass::runOnOperation() {
+  func::FuncOp funcOp = getOperation();
+  int64_t useSize = size.getValue();
+  uint64_t unroll = unrollFactor.getValue();
+  if (unroll == 0) unroll = 1;
```
If you always want `1` when it's zero, why not set the default value to `1`?

vmurali replied:
If the user sets it to 0 by mistake, I change it back to 1.

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -366,6 +411,128 @@ void LinalgFusePass::runOnOperation() {
+    doSplitReduction =
+        getSplitReductionOptionsFromConfig(funcOp, useSize, unroll);
+  }
+  if (doSplitReduction) {
```
Let's return directly if doSplitReduction is false. It can reduce the indents.

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -366,6 +411,128 @@ void LinalgFusePass::runOnOperation() {
+  uint64_t unroll = unrollFactor.getValue();
+  if (unroll == 0) unroll = 1;
+  bool doSplitReduction = true;
+  if (useSize == 0) {
```
Let's return directly if `useSize` is zero.

vmurali replied:
It's not returning - it's the default, and I should run `getSplitReductionOptionsFromConfig` in that case.

hanhanW commented on tests/e2e/regression/split_reduction.mlir:
```diff
@@ -0,0 +1,73 @@
+#map0 = affine_map<(d1,d2) -> (d1,d2)>
```
let's move the file to `tests/e2e/cpu_specific/`

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/Passes.td:
```diff
@@ -67,6 +67,22 @@ def LinalgFuse : Pass<"linalg-fuse", "func::FuncOp"> {
+  ];
+  let options = [
+    // This corresponds to the value in the last level of tiling
+    Option<"size", "size", "int64_t", /*default=*/"0",
+      "Size that splitReduction should use for parallel reduction results.">,
```
IMO, we can name it `ratio`. That's the naming from upstream.

vmurali replied:
Ratio seems a completely random name, even if upstream uses it. What I need is the final size of the parallel dimension. I would strongly favor leaving it as is.

vmurali replied:
Size=0 triggers reading the lowering_config and if there's no lowering_config, the pass doesn't execute. Executing the pass with tile size of 1 is strictly worse than not executing it (and instead using the default linalg_tile_and_fuse)

hanhanW replied:
Thanks for the explanation, please document it in the td file as well, thanks!

pzread commented:
@vmurali After you push the next commit, if the benchmarks don't run, can you try to remove the labels and add them again?

dcaballe reviewed (changes requested):
Looks good! Some comments!

vmurali commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -267,7 +312,7 @@ void LinalgFusePass::runOnOperation() {
       getTileAndFuseOptionsFromConfig(funcOp, tilingLevel);
   if (failed(defaultTilingOptions)) {
-    return signalPassFailure();
+    return;
```
It seems to create problems downstream. I am guessing that signalPassFailure() kills the entire pipeline, not just this pass.

vmurali commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -366,6 +411,128 @@ void LinalgFusePass::runOnOperation() {
+    // Make sure that only generic ops are supported, only one input and the
+    // innermost tensor dimension is reduction dimension.  Everything else is
+    // copied from the original SplitReduction Transformation.
+    if (failed(filter.checkAndNotify(rewriter, op)) ||
+        !op.hasTensorSemantics() || op.getNumReductionLoops() != 1 ||
+        op.getNumOutputs() != 1 || !op.hasOnlyProjectedPermutations() ||
+        !isa<linalg::GenericOp>(op) || op.getNumInputs() != 1)
+      return rewriter.notifyMatchFailure(op, "precondition not met");
```
Actually the only thing that has to be checked is if it has innermost reduction dimension (for all inputs). I can remove the rest of the checks and it should still work. The other checks are not present in SplitReduction (single input operand, single reduction loop, tensor semantics, single output, projected permutations)

vmurali replied:
SplitReduction already requires a bunch of these checks to function correctly. The only extra check is `isa<Linalg::GenericOp>(op)  && op.getNumInputs() == 1`. Both of these can be removed. If I remove getNumInputs, I should modify last Dimension check to span all inputs. I can add a todo for now.

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -138,6 +140,40 @@ getTileAndFuseOptionsFromConfig(func::FuncOp funcOp, int64_t tilingLevel) {
+  auto parallelSizesAttr = tileSizesAttr[tileSizesAttr.size() - 2];
+  if (!parallelSizesAttr) return false;
+  auto parallelSizes = parallelSizesAttr.cast<ArrayAttr>();
+  if (!parallelSizes || parallelSizes.empty()) return false;
```
Yes, that's IREE's coding convention.

vmurali commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -267,7 +312,7 @@ void LinalgFusePass::runOnOperation() {
       getTileAndFuseOptionsFromConfig(funcOp, tilingLevel);
   if (failed(defaultTilingOptions)) {
-    return signalPassFailure();
+    return;
```
It seems to create problems downstream. I am guessing that signalPassFailure() kills the entire pipeline, not just this pass.

vmurali replied:
`getTileAndFuseOptionsFromConfig` also fails if there's no LoweringConfig defined in the vicinity (I don't know what the vicnity is, but when I remove the `lowering_config` from the op, it fails; I want it to not fail then). (It looks like you have assumed `getTileAndFuseOptionsFromConfig` fails only when tiling_level is `-1`.)

hanhanW replied:
We use the lowering_config to drive all sandbox fine-grained transforms. After exploring all the code and logic, I think I understand what's happening here. The two generic ops created by SplitReduction, and might both of them carry the configuration. While getRootOp assumes that only one config can be existed, it return a failure().

IMO, silently skipping the pass is confusing. To address the issue, we should either

1) Fix the getRootOp to check if the configs are identical. It might raise other issues because you always run tiling on reductions right after SplitReduction.
2) Signal something when skipping the pass.

Though the fix hides the actual issue, I'm fine with the current change, since I can see the CodegenDriver stuff will be deprecated maybe in the near future. I'm just writing down what I found in this change.

vmurali commented:
It looks like I am still not able to generate the benchmark tables after running the checks @pzread, @hanhanW

pzread commented:
Looks like they are running now
https://buildkite.com/iree/iree-benchmark-android/builds/4411
https://buildkite.com/iree/iree-benchmark-linux/builds/1630

vmurali commented:
Performance results:
https://gist.github.com/iree-github-actions-bot/09a2bb55a134d210206d0d3c4abf01ff

dcaballe commented:
Wow, 8-10% improvement is a lot, taking into account that this only triggers for reductions in generic ops! Well done!
I would expect the only regression on Pixel to be noise? It's also suspicious that we don't see changes in RVV dispatch sizes. I would expect some variation there. I'll give it a try once integration lands.

vmurali commented:
> Wow, 8-10% improvement is a lot, taking into account that this only triggers for reductions in generic ops! Well done! I would expect the only regression on Pixel to be noise? It's also suspicious that we don't see changes in RVV dispatch sizes. I would expect some variation there. I'll give it a try once integration lands.

I just added the android one (the previous two were both linux). Not sure if linux+Pixel is run in the linux benchmark.

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -140,6 +142,41 @@ getTileAndFuseOptionsFromConfig(func::FuncOp funcOp, int64_t tilingLevel) {
+  }
+  unsigned numTileLevels =
+      mlir::iree_compiler::getNumTileLevels(rootOp.value());
+  assert(numTileLevels >= 2 && "at least 2 tiling levels must be present");
```
Seems like an unnecessary check.

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -368,6 +419,135 @@ void LinalgFusePass::runOnOperation() {
+    auto optionsFirst =
+        linalg::LinalgTilingOptions().setTileSizes(tileSizesSVFirst);
+    FailureOr<linalg::TiledLinalgOp> tileResFirst =
+        linalg::tileLinalgOp(rewriter, op, optionsFirst);
```
Note that this is being deprecated. You should be able to use `scf::tileUsingSCFForOp` method in `mlir/Dialect/SCF/Transforms/TileUsingInterface.h`

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -368,6 +419,135 @@ void LinalgFusePass::runOnOperation() {
+      : OpInterfaceRewritePattern<linalg::LinalgOp>(context, benefit),
+        reductionReordering(reductionReordering),
+        size(size),
+        unrollFactor(unrollFactor),
```
It seems like `unrollFactor` isnt used in this PR. I'd drop it for now.

MaheshRavishankar reviewed (changes requested):
Great! Mostly looks fine! I am not sure why the `unrollFactor` is being added in this PR. Apart from that looks good!

dcaballe commented on compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:
```diff
@@ -52,6 +52,12 @@ static llvm::cl::opt<bool> clEnableMicrokernels(
+    "iree-llvmcpu-enable-reduction-reordering",
+    llvm::cl::desc("Enables reordering reductions for floats that are not"
+                   "(strictly) associative for addition or multiplication"),
+    llvm::cl::init(true));
```
I thought the idea was to enable reassociation by default, didn't it?

vmurali replied:
Hmm, the default should be to take this path right? Why should the default be false?

hanhanW replied:
ah yes, I missed that we've checked the element types should be integers. In the integer case, it should be default on. Sorry about that.

For the floating point, it should be default off, and have a flag to control it right?

hanhanW replied:
Can we resolve the conversation only when there are no other questions or the issue is addressed... I still don't get the answer about FP yet...

I do have concerns if it's default on for FP. I did spend couple weeks on debugging numerical issues because of a similar feature. IMO, we should not enable the features for FP by default and should have a flag for controlling it. People who care about performance (more than numerical) can enable the feature themselves. LLVM also has the same restriction. It's off by default, and it's on in fast-math mode. https://clang.llvm.org/docs/UsersManual.html#id12

I assumed that the flag is for controlling the behavior for FP, but it seems not to me now.. It's very confusing to me because the description is about enabling the features for FP or not.

vmurali replied:
Okay I will let you resolve the conversation from now on. I just use resolve whenever I think it's resolved or I have made code changes to fix the issue usually.

> I do have concerns if it's default on for FP. I did spend couple weeks on debugging numerical issues because of a similar feature. IMO, we should not enable the features for FP by default and should have a flag for controlling it. People who care about performance (more than numerical) can enable the feature themselves. LLVM also has the same restriction. It's off by default, and it's on in fast-math mode. https://clang.llvm.org/docs/UsersManual.html#id12

Okay. I will make it off by default.

> I assumed that the flag is for controlling the behavior for FP, but it seems not to me now.. It's very confusing to me because the description is about enabling the features for FP or not.

Yes it is. The feature is turned on by default for all integer types. For any other type (FP included) it is controlled by the flag

dcaballe replied:
Yes, let's rename this flag to `iree-llvmcpu-enable-fp-reduction-reassoc` or similar.

Regarding its default value, I'm ok with setting it up to `false` by default if there are concerns. However, we have to be consistent. Are we using precise implementations for all the arithmetic and math operations subject to leading to numerical differences (e.g., div, rem, sqrt, exp, log, etc.? What are the reference TFLite implementations doing? Perhaps this is something to bring up with the rest of the team.

hanhanW replied:
> Are we using precise implementations for all the arithmetic and math operations subject to leading to numerical differences (e.g., div, rem, sqrt, exp, log, etc.?

Yes, we are.

> What are the reference TFLite implementations doing? Perhaps this is something to bring up with the rest of the team.

I don't know what TFLite implements is. I know that there are many flags in reference tensorflow implementation. When doing correctness tests, we disable aggressive flags. (one of the examples: https://github.com/iree-org/iree/issues/9853#issuecomment-1191913120)

My main point is that given a dispatch or a model, all the backends in IREE should generate identical results by default (for FP cases, the accuracy could be <= 1e-6). Turning the feature on by default will make IREE backends get diverged. The fancy optimization should be controlled under a flag (say iree-fast-math) if they could impact accuracy a lot.

vmurali replied:
Added a flag `--iree-llvmcpu-enable-reduction-reordering-for-all`

ThomasRaoux replied:
> My main point is that given a dispatch or a model, all the backends in IREE should generate identical results by default (for FP cases, the accuracy could be <= 1e-6). Turning the feature on by default will make IREE backends get diverged. The fancy optimization should be controlled under a flag (say iree-fast-math) if they could impact accuracy a lot.

I agree with the idea. I just want to point out that this isn't how it is on the GPU backends (both llvmgpu and spirv) right now as not distributing reduction makes performance collapse. This is also true for most framework out there. I agree that this is unfortunate and it would be nice to have a mode with strict floating points but I don't think we will be able to afford to not do this kind of re-association for reductions by default.

vmurali replied:
> I just want to point out that this isn't how it is on the GPU backends (both llvmgpu and spirv) right now as not distributing reduction makes performance collapse. This is also true for most framework out there. 

Are you implying that it would diverge more if this is not turned on?

ThomasRaoux replied:
not really, I just meant this is necessary for some targets. Every target is likely to do partial accumulation with a different tile size so we can't really bet on all the target doing re-association and expect the same result.

hanhanW replied:
> I doubt we can diverge that much by reordering reductions. But nevertheless I have made the flag false by default.

Here is an example that the difference increased from 1e-6 to 1e-2 because of re-association optimization. https://github.com/iree-org/iree/issues/9536

We could see more examples when connecting more models e2e.

I'm surprised that CPU and GPU backends are diverged today. :o 
IMO, we should control these fancy optimizations under one flag. I can't speak if they will be on by default in the future, but having them under one flag controlled and having ability to generate "golden" values are important. It's also easier if we really want to flip to that pass by default in the future.

vmurali replied:
> Here is an example that the difference increased from 1e-6 to 1e-2 because of re-association optimization. https://github.com/iree-org/iree/issues/9536

Interesting. Looks like there's a F64 to F32 conversion as well. Is that orthogonal to the rounding bug?

ThomasRaoux replied:
> I'm surprised that CPU and GPU backends are diverged today. :o
> IMO, we should control these fancy optimizations under one flag. I can't speak if they will be on by default in the future, but having them under one flag controlled and having ability to generate "golden" values are important. It's also easier if we really want to flip to that pass by default in the future.

I completely agree there should be a flag but it would have to be a per operation flag to be meaningful. I believe Jakub has started some discussion on discourse about fast math but this is going to be a non trivial change.

> Interesting. Looks like there's a F64 to F32 conversion as well. Is that orthogonal to the rounding bug?

I think this is quite different.converting F64 to F32 is usually plain wrong.

hanhanW replied:
Thomas, thanks for the clarification, it really help!

> Interesting. Looks like there's a F64 to F32 conversion as well. Is that orthogonal to the rounding bug?

That is a different issue. The root cause is re-association. https://github.com/iree-org/iree/issues/9536#issuecomment-1164844079

Anyway, it's just an example issue that I've hit about re-association in IREE. I'm happy to see the pass on for int by default, and off for FP now. Thanks for pushing this forward, it's an awesome work!

dcaballe replied:
Yeah, reduction error can surprisingly accumulate and lead to major output differences. I agree with the idea of having a higher precision mode but I'm not sure if the bit-exact goal across targets is feasible. Different targets have different instructions for fundamental operations using different approximations which would lead to mismatching precision. We would have to use the same high-accuracy software-based emulation to get something along those lines. Not sure how realistic this is.

vmurali replied:
Can we resolve this comment?

hanhanW replied:
The issue is addressed, but I'd like to keep the conversation open. I found that keeping this kind of conversation open is helpful when people play something similar and could lead them to the PR to see some discussions.

dcaballe replied:
Just my opinion here... I don't like to resolve the comments because it hides the discussion and it's difficult for reviewers to check if the comments were addressed or not. As a reviewer, you have to go and open all the comments one by one. To mark comments as addressed for myself, I use an emoji and that usually works well for me. Of course, you can do whatever works better for you :)

dcaballe commented on compiler/src/iree/compiler/Codegen/Dialect/LoweringConfig.cpp:
```diff
@@ -294,6 +294,11 @@ SmallVector<Value, 4> getTileSizes(OpBuilder &b, Operation *op,
         return b.create<arith::ConstantIndexOp>(op->getLoc(), t);
       }));
 }
+unsigned getNumTileLevels(Operation *op) {
```
nit: while line

dcaballe commented on compiler/src/iree/compiler/Codegen/Dialect/LoweringConfig.h:
```diff
@@ -107,6 +107,7 @@ IREE::Codegen::LoweringConfigAttr getLoweringConfig(Operation *op);
 /// `iree_codegen.lowering_config` attribute is set on it.
 SmallVector<int64_t> getTileSizes(Operation *op, unsigned level);
 SmallVector<Value, 4> getTileSizes(OpBuilder &b, Operation *op, unsigned level);
+unsigned getNumTileLevels(Operation *op);
```
nit: white line + doc as this is public API.

vmurali commented on compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir:
```diff
@@ -59,7 +59,10 @@ hal.executable private @check_no_cse {
 //  CHECK-NOT:    memref.alloc
 //      CHECK:    %[[FOR:.+]] = scf.for
-//      CHECK:    %[[DIVF:.+]] = arith.divf %[[FOR]]
+//      CHECK:    %[[EXTRACT:.+]] = vector.extract %[[FOR]]
+//      CHECK:    %[[REDUCTION:.+]] = vector.reduction <add>, %[[EXTRACT]]
+//      CHECK:    %[[INS:.+]] = vector.insertelement %[[REDUCTION]]
+//      CHECK:    %[[DIVF:.+]] = arith.divf %[[INS]]
```
Referring to above comment (default should be true).

dcaballe commented on compiler/src/iree/compiler/Codegen/LLVMCPU/test/split_reduction_pipeline_tests.mlir:
```diff
@@ -0,0 +1,281 @@
+// CHECK:         scf.for
+// CHECK:           scf.for
+// CHECK:             scf.for %[[ARG3:.+]] = %[[C0]] to %[[C64]] step %[[C1]]
+// CHECK:               scf.yield %{{.+}} : vector<1x1x4xi32>
```
match the add operation also?

dcaballe commented on compiler/src/iree/compiler/Codegen/LLVMCPU/test/split_reduction_pipeline_tests.mlir:
```diff
@@ -0,0 +1,281 @@
+      hal.return %x, %y, %z : index, index, index
+    }
+    builtin.module {
+      func.func @split_reduction_pass2_dispatch_0_generic_1024xDx256() {
```
+1, using a name that indicates what this is testing (inner_red, outer_red, fuse, etc.) and optionally adding a small comment about the supported or not supported case that is being tested is very useful.

vmurali replied:
`fail` indicates it's not supported. I will change the rest of the names, and also remove the dimension names.

vmurali commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -368,6 +419,135 @@ void LinalgFusePass::runOnOperation() {
+
+  LogicalResult matchAndRewrite(linalg::LinalgOp op,
+                                PatternRewriter &rewriter) const override {
+    // Make sure that only generic ops are supported, only one input and the
+    // innermost tensor dimension is reduction dimension.  Everything else is
+    // copied from the original SplitReduction Transformation.
+    // TODO: support named ops, numInputs > 1, and modify lastDim check below
+    // accordingly. If reductionReordering is not enabled by default, it must be
+    // an integer or index type to proceed to allow associative reordering.
+    if (failed(filter.checkAndNotify(rewriter, op)) ||
+        !op.hasTensorSemantics() || op.getNumReductionLoops() != 1 ||
+        op.getNumOutputs() != 1 || !op.hasOnlyProjectedPermutations() ||
+        !isa<linalg::GenericOp>(op) || op.getNumInputs() != 1 ||
+        !(reductionReordering || op.getInputOperand(0)
+                                     ->get()
+                                     .getType()
+                                     .dyn_cast<ShapedType>()
+                                     .getElementType()
+                                     .isIntOrIndex()))
```
Take a look at the current splitReduction code here. https://github.com/llvm/llvm-project/blob/81ca5aa452400843235e058bc9c83fe71eccd593/mlir/lib/Dialect/Linalg/Transforms/SplitReduction.cpp#L65

I am exposing exactly those conditions + 3 more.

vmurali replied:
Exposing the SplitReduction pass's preconditions is useless now because I am going to remove most of those preconditions in my next change to splitReduction - I don't want to muck around with LLVM changes again because it looks like it will take an enormous amount of time to get that synced to IREE (given last week's experience).

vmurali replied:
Some of these preconditions are necessary for the IREE pass also: ( !op.hasTensorSemantics() || op.getNumReductionLoops() != 1 ||
        op.getNumOutputs() != 1 || !op.hasOnlyProjectedPermutations())

Even if these preconditions are supported by the base SplitReduction, we are not going to support it in IREE. I can adjust the comment to reflect that we need them.

vmurali commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -186,6 +223,20 @@ struct LinalgFusePass : public LinalgFuseBase<LinalgFusePass> {
+  void runOnOperation() override;
+
+ private:
+  bool reductionReordering = true;
```
We can't because reductionReordering is *always* enabled if the operands of the pass is integers. It's not enabled *only if* the operands are floats or other data types where reductionReordering can lead to different results

vmurali replied:
Can we resolve this comment?

MaheshRavishankar replied:
I think having both is fine. Unfortunately we need both pipeline level flag and pass level flag (latter at least for testing). It makes sense to use the pipeline level flag to set the flag of the pass. This is what MLIR is moving towards as well.

Ill resolve this comment.

hanhanW replied:
Thanks for the explanation, I now understand what we want to achieve. Can we rename it to reductionFPReordering? The integer cases are always reordered in the pass, and the FP cases are reordered only when the flag is true?

vmurali replied:
> Can we rename it to reductionFPReordering?
It's called clEnableReassociateFpReductions based on what @dcaballe suggested.

> The integer cases are always reordered in the pass, and the FP cases are reordered only when the flag is true?
Yes

vmurali commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -368,6 +419,135 @@ void LinalgFusePass::runOnOperation() {
+                                     .getElementType()
+                                     .isIntOrIndex()))
+      return rewriter.notifyMatchFailure(op, "precondition not met");
+    SmallVector<unsigned> dims;
+    op.getReductionDims(dims);
+    AffineMap map = op.getMatchingIndexingMap(op.getInputOperand(0));
+    unsigned lastIdx = map.getNumResults() - 1;
+    unsigned lastDim = map.getDimPosition(lastIdx);
+    if (lastDim != dims[0]) {
+      return rewriter.notifyMatchFailure(op, "precondition not met");
+    }
+
```
> My understanding is that we'd like to check if the last dimension is reduction. We should derive the information from the Linalg op generally because it's how structure ops (i.e., linalg ops) work. We don't need to build the condition based on the precondition of hasOneInput, hasExactlyOneReduction, etc.

No, we have to check that the last rank is reduction dimension.

dcaballe commented on compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:
```diff
@@ -52,6 +52,12 @@ static llvm::cl::opt<bool> clEnableMicrokernels(
     llvm::cl::init(false));
 
+static llvm::cl::opt<bool> clEnableReductionReorderingForAll(
+    "iree-llvmcpu-enable-reduction-reordering-for-all",
```
Would suggest being explicit about `fp` in the flag.

vmurali replied:
So, the idea is to keep it generic so that if we support more types (like fixed points, complex, etc), we will still rely on this flag to allow reduction reordering for operations on those types.

dcaballe replied:
I would stick to YAGNI and change the flag in the future if at all needed for fixed point or complex but if you like the existing one, we should at least specify what `for-all` means (all codegens, all types, etc.?).

Even better, we could align the flag with MLIR and LLVM:
```
mlir-opt --help | grep fp
        --reassociate-fp-reductions                      - Allows llvm to reassociate floating-point reductions for speed
```
Anyways, I'm bikeshedding here. Don't have a strong opinion. I'll leave it to you.

dcaballe commented on compiler/src/iree/compiler/Codegen/LLVMCPU/test/pipeline_tests.mlir:
```diff
@@ -63,6 +64,16 @@ hal.executable private @check_no_cse {
+//      REORDERCHECK:    %[[DIVF:.+]] = arith.divf %[[INS]]
+//      REORDERCHECK:    %[[RES:.+]] = vector.extract %[[DIVF]]
+//      REORDERCHECK:    memref.store %[[RES]]
+
```
Hmm... why don't we move this test to the split_reduction test below and leave this file intact? This file seems to be testing something unrelated to reductions and we only have `REORDERCHECK` in one of the tests.

dcaballe commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -368,6 +405,138 @@ void LinalgFusePass::runOnOperation() {
+
+  LogicalResult matchAndRewrite(linalg::LinalgOp op,
+                                PatternRewriter &rewriter) const override {
+    // Make sure that
+    // a) The pass was not applied earlier
+    // b) operation has tensor semantics
+    // c) number of reduction loops == 1
+    // d) operation only has projection permutations for index map
+    // e) It's a linalg generic op
+    // f) it has only 1 input
+    // g) if enableReductionReordering is not set, then operand is an int
+    // h) reduction dimension is the innermost dimension of the operand shape
+    // TODO: support named ops, numInputs > 1, and modify lastDim check below
+    // accordingly. If reductionReordering is not enabled by default, it must be
+    // an integer or index type to proceed to allow associative reordering.
+    if (failed(filter.checkAndNotify(rewriter, op)) ||
+        !op.hasTensorSemantics() || op.getNumReductionLoops() != 1 ||
+        op.getNumOutputs() != 1 || !op.hasOnlyProjectedPermutations() ||
+        !isa<linalg::GenericOp>(op) || op.getNumInputs() != 1 ||
+        !(reductionReordering || op.getInputOperand(0)
+                                     ->get()
+                                     .getType()
+                                     .dyn_cast<ShapedType>()
+                                     .getElementType()
+                                     .isIntOrIndex()))
+      return rewriter.notifyMatchFailure(
+          op,
+          "pass was applied earlier or "
+          "operation doesn't have tensor semantics or "
+          "number of reduction loops != 1 or "
+          "operation has non projected permutations for index map or "
+          "it's not a linalg generic op or "
+          "it has multiple inputs or "
+          "if enableReductionReordering is not set and "
+          "operand is not an int");
```
Please, split these checks into individual ones so that the condition per check is simpler and we can return a single specific error message.

dcaballe reviewed (approved):
Thanks for addressing all the feedback. I added a few more comments. LGTM once you address those! Thanks

vmurali commented:
> Thanks for addressing all the feedback. I added a few more comments. LGTM once you address those! Thanks

Thanks for the LGTM. I have made the changes that you had asked (except for renaming the option). I will wait till you and Jakub finish integration of llvm-fork before I push this.

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -368,6 +405,150 @@ void LinalgFusePass::runOnOperation() {
+    if (op.getNumInputs() != 1) {
+      return rewriter.notifyMatchFailure(op, "doesn't have exactly 1 input");
+    }
+    if (!(reductionReordering || op.getInputOperand(0)
```
Nit: Might be easier to just do
```
auto elementType = op.getInputOperand(0)...getElementType()
if (!(reductionReording || elementType.isIntOrIndex()) {...}
```

vmurali replied:
Hmm, not sure what that buys us; anyways, ignoring the nit till we need CSE of `elementType` for something else.

MaheshRavishankar reviewed (dismissed):
PR itself is fine. Could you split the PR to have the cherry pick separate.

vmurali commented:
> PR itself is fine. Could you split the PR to have the cherry pick separate.

I don't think there's any cherry pick here.

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -368,6 +405,150 @@ void LinalgFusePass::runOnOperation() {
+    if (op.getNumOutputs() != 1) {
+      return rewriter.notifyMatchFailure(op, "doesn't have exactly 1 output");
+    }
+    if (!op.hasOnlyProjectedPermutations()) {
+      return rewriter.notifyMatchFailure(
+          op, "index map doesn't have only projected permutations");
+    }
```
I don't really understand why we can't reuse the upstream preconditions in SplitReduction here, and have to do same checks in advance. E.g., hasOnlyProjectedPermutations looks like the needs from SplitReduction pass, not IREE specific reason. Maybe there are deep reasons that I could not figure out. I'm fine with change, and will leave the decision to you.

vmurali replied:
That is perhaps the only check that is necessary for `splitReduction` to work. Might as well perform the check here instead of creating a precondition checker in llvm.

hanhanW replied:
The `hasTensorSemantics`, `getNumReductionLoops`, `getNumOutputs` are all requirements from upstream SplitReduction. filter stuff is fine because that's the way to drive CodegenDriver.

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -368,6 +405,150 @@ void LinalgFusePass::runOnOperation() {
+  LogicalResult matchAndRewrite(linalg::LinalgOp op,
+                                PatternRewriter &rewriter) const override {
+    // Make sure that
+    // - the pass has not been applied before
+    // - has tensor semantics
+    // - number of reduction loops == 1
+    // - has exactly 1 output
+    // - index map has only projected permutations
+    // - is a linalg generic op
+    // - has exactly 1 input
+    // - if enableReductionReordering is not set, then operand is an int
+    // - innermost dimension of the input operand is reduction
```
[optional] the failure messages already tells us the check. The comments are redundant to me. I do know some people like to write such comments, so I'm not marking the change required. Just an FYI that I've seen style guide discussions in Google that asking people don't write redundant comments. The best is to make them self-describing.

vmurali replied:
I needed the comment to write the code to not lose track of all the checks.

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -186,6 +210,19 @@ struct LinalgFusePass : public LinalgFuseBase<LinalgFusePass> {
+struct LinalgSplitReductionPass
+    : public LinalgSplitReductionBase<LinalgSplitReductionPass> {
+  LinalgSplitReductionPass(bool enableReductionReordering, int64_t size = 0,
+                           uint64_t unrollFactor = 1) {
```
unrollFactor is not used. It seems unrelated to the PR. How about removing it and reintroducing when there's a need?

hanhanW reviewed (changes requested):
Can you exclude the changes in llvm-project or split it out?

MaheshRavishankar commented:
> > PR itself is fine. Could you split the PR to have the cherry pick separate.
> 
> I don't think there's any cherry pick here.

The changes to LLVM project need to go into a separate PR?

vmurali commented:
> > > PR itself is fine. Could you split the PR to have the cherry pick separate.
> > 
> > 
> > I don't think there's any cherry pick here.
> 
> The changes to LLVM project need to go into a separate PR?

I am not going to push this PR until integrate happens (and that includes my PR in LLVM), @hanhanW

vmurali commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -140,6 +143,27 @@ getTileAndFuseOptionsFromConfig(func::FuncOp funcOp, int64_t tilingLevel) {
 
+/// Default method to initialize the split reduction size in IREE. These could
+/// be overriden by the command line options if specified.
+static FailureOr<int64_t> getSplitReductionOptionsFromConfig(
+    func::FuncOp funcOp) {
```
Once I have unroll, I will have to change it again. Kept it as `Options` for future-proofing.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/test/split_reduction_pipeline_tests.mlir:
```diff
@@ -0,0 +1,348 @@
+// REORDERCHECK:           arith.addf %{{.+}}, %{{.+}} : vector<1x4xf32>
+
+// -----
+
+
```
delete dup blank line

```suggestion

```

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/test/split_reduction_pipeline_tests.mlir:
```diff
@@ -0,0 +1,348 @@
+// RUN: iree-opt --pass-pipeline='hal.executable(hal.executable.variant(iree-llvmcpu-lower-executable-target))' --split-input-file %s | FileCheck %s
```
Ideally we should put the test to https://github.com/iree-org/iree/tree/main/compiler/src/iree/compiler/Codegen/Sandbox/test and test with the pass itself.

I'm fine with this because the tests are covered.

hanhanW reviewed (approved):
Thanks for addressing the comments. Just one nit on the variable name and function default argument.

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/Passes.h:
```diff
@@ -38,6 +38,10 @@ std::unique_ptr<OperationPass<func::FuncOp>> createLinalgFusePass();
 
+/// Creates a pass to perform SplitReduction transformations of `LinalgOp`s.
+std::unique_ptr<OperationPass<func::FuncOp>> createLinalgSplitReductionPass(
+    const bool enableReductionReordering = true, const int64_t size = 0);
```
I see the reason now. Why not do something like other passes (i.e., declare a function that takes zero arguments and implement it)? Is it for cleaner? I don't have a strong opinion on it, will leave the decision to you.

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/LinalgTensorCodegenDriver.cpp:
```diff
@@ -186,6 +208,18 @@ struct LinalgFusePass : public LinalgFuseBase<LinalgFusePass> {
+  void runOnOperation() override;
+
+ private:
+  bool fpReductionReordering = true;
```
just found that this should be false by default for consistency.

vmurali commented:
https://github.com/iree-org/iree/issues/9981

hanhanW commented on compiler/src/iree/compiler/Codegen/Sandbox/Passes.h:
```diff
@@ -38,6 +38,10 @@ std::unique_ptr<OperationPass<func::FuncOp>> createLinalgFusePass();
 
+/// Creates a pass to perform SplitReduction transformations of `LinalgOp`s.
+std::unique_ptr<OperationPass<func::FuncOp>> createLinalgSplitReductionPass(
+    const bool enableReassociateFpReductions = false, const int64_t size = 0);
```
please drop meaningless `const`. (a) it's not consistent with others. (b) They are meaningless in function declaration. See https://abseil.io/tips/109 for more details.

```suggestion
   bool enableReassociateFpReductions = false, int64_t size = 0);
```

vmurali replied:
Done
